xdem/spstats.py

```diff
@@ -0,0 +1,702 @@
+    """
+    Covariance function based on variogram function (COV = STD - VGM)
+
+    :param h: spatial lag
+    :param crange: correlation range
+    :param model: model
+    :param psill: partial sill
+    :param kappa: smoothing parameter for Exp Class
+    :param nugget: nugget
+
+    :returns: covariance function
+    """
+
+    return (nugget + psill) - vgm(h,crange,model=model,psill=psill,kappa=kappa)
+
+def vgm(h: float, crange: float, model: str = 'Sph', psill: float = 1., kappa: float = 1/2, nugget: float = 0):
+    """
+    Compute variogram model function (Spherical, Exponential, Gaussian or Exponential Class)
+
+    :param h: spatial lag
+    :param crange: correlation range
+    :param model: model
+    :param psill: partial sill
+    :param kappa: smoothing parameter for Exp Class
+    :param nugget: nugget
+
+    :returns: variogram function
+    """
+
+    c0 = nugget #nugget
+    c1 = psill #partial sill
+    a1 = crange #correlation range
+    s = kappa #smoothness parameter for Matern class
+
+    if model == 'Sph':  # spherical model
+        if h < a1:
+            vgm = c0 + c1 * (3 / 2 * h / a1-1 / 2 * (h / a1) ** 3)
+        else:
+            vgm = c0 + c1
+    elif model == 'Exp':  # exponential model
+        vgm = c0 + c1 * (1-np.exp(-h / a1))
+    elif model == 'Gau':  # gaussian model
+        vgm = c0 + c1 * (1-np.exp(- (h / a1) ** 2))
+    elif model == 'Exc':  # stable exponential model
+        vgm = c0 + c1 * (1-np.exp(-(h/ a1)**s))
+
+    return vgm
+
+def std_err_finite(std: float, neff_tot: float, neff: float) -> float:
+    """
+    Standard error of subsample of a finite ensemble
+
+    :param std: standard deviation
+    :param neff_tot: maximum number of effective samples
+    :param neff: number of effective samples
+
+    :return: standard error
+    """
+    return std * np.sqrt(1 / neff_tot * (neff_tot - neff) / neff_tot)
+
+def std_err(std: float, neff: float) -> float:
+    """
+    Standard error
+
+    :param std: standard deviation
+    :param neff: number of effective samples
+
+    :return: standard error
+    """
+    return std * np.sqrt(1 / neff)
+
+def distance_latlon(tup1: tuple, tup2: tuple) -> float:
+    """
+    Distance between two lat/lon coordinates projected on a spheroid
+    :param tup1: lon/lat coordinates of first point
+    :param tup2: lon/lat coordinates of second point
+
+    :return: distance
+    """
+
+    # approximate radius of earth in km
+    R = 6373000
+
+    lat1 = m.radians(abs(tup1[1]))
+    lon1 = m.radians(abs(tup1[0]))
+    lat2 = m.radians(abs(tup2[1]))
+    lon2 = m.radians(abs(tup2[0]))
+
+    dlon = lon2 - lon1
+    dlat = lat2 - lat1
+
+    a = m.sin(dlat / 2)**2 + m.cos(lat1) * m.cos(lat2) * m.sin(dlon / 2)**2
+    c = 2 * m.atan2(m.sqrt(a), m.sqrt(1 - a))
+
+    distance = R * c
+
+    return distance
+
+def kernel_sph(xi: float, x0: float, a1: float) -> float:
```

Comment:
This is just an idea, but could `@np.vectorize` be used to speed up some calculations significantly?

The decorator allows for arrays to be passed to the function directly and does element-wise operations in near-C speed.

Maybe we can try it in a future PR!

Reply:
Added issue #40